### ISSUE TITLE
Does it monitor AWS ECR ?

### ISSUE THREAD

**Issue by siredwin** [fmartinou/whats-up-docker#12]
I have given up on watchtower. 
I think I need something simpler and slightly manual for production.
Will this monitor ECR repo? I am also guessing I will see the full information through bash?
I will be testing this later today or early tomorrow.

**Comment by siredwin** [comment]
As I suspected, it seems not. I am seeing errors

**Comment by fmartinou** [comment]
Hi,

Currently, I've only tested the app with the official Docker Hub.
Because ECR is using the same V2 API than the Docker hub, I guess that it should be working with some minor adjustments.
I study your request as soon as I find some free time.
In the meantime, can you send me the tests you did and the errors you got?

**Comment by siredwin** [comment]
This is the error:

```
{"name":"whats-up-doc","hostname":"wud","pid":1,"level":20,"err":{"message":"Error: Invalid URI \"account.dkr.ecr.us-east-2.amazonaws.com/v2/repositories/library/gotify/tags?page=1\"","name":"RequestError","stack":"RequestError: Error: Invalid URI \"account.dkr.ecr.us-east-2.amazonaws.com/v2/repositories/library/gotify/tags?page=1\"\n    at new RequestError (/home/node/app/node_modules/request-promise-core/lib/errors.js:14:15)\n    at Request.plumbing.callback (/home/node/app/node_modules/request-promise-core/lib/plumbing.js:87:29)\n    at Request.RP$callback [as _callback] (/home/node/app/node_modules/request-promise-core/lib/plumbing.js:46:31)\n    at self.callback (/home/node/app/node_modules/request/request.js:185:22)\n    at Request.emit (events.js:314:20)\n    at Request.init (/home/node/app/node_modules/request/request.js:273:17)\n    at Request.RP$initInterceptor [as init] (/home/node/app/node_modules/request-promise-core/configure/request2.js:45:29)\n    at new Request (/home/node/app/node_modules/request/request.js:127:8)\n    at request (/home/node/app/node_modules/request/index.js:53:10)\n    at getTagsPage (/home/node/app/watcher/providers/docker/Docker.js:19:12)"},"msg":"Error: Invalid URI \"account.dkr.ecr.us-east-2.amazonaws.com/v2/repositories/library/gotify/tags?page=1\"","time":"2020-08-12T04:10:50.250Z","v":0}
```

**Comment by fmartinou** [comment]
ECR is a service hosting **private** registries so I have to figure out how to deal with credentials first:
- [HTTP API Authentication](https://docs.aws.amazon.com/AmazonECR/latest/userguide/Registries.html)
- [Get authorization token](https://docs.aws.amazon.com/cli/latest/reference/ecr/get-authorization-token.html#:~:text=Authorization%20tokens%20are%20valid%20for%2012%20hours.&text=The%20registry%20URL%20to%20use,1.amazonaws.com%20..)

**Comment by siredwin** [comment]
I am not able to contribute since I am not good with JavaScript. But I believe your service will be beneficial in production environments.

**Comment by fmartinou** [comment]
I perfectly agree.
Supporting other registries like [quay](https://quay.io/), [gcr](https://cloud.google.com/container-registry), [ecr](https://aws.amazon.com/ecr/)... would be a big improvement.

I just need to find some time to add them one after another.

**Comment by siredwin** [comment]
Is there a reason why you chose JavaScript for this?

**Comment by fmartinou** [comment]
I really ❤️ this language (both frontend and backend) 😍 😍 😍 
With JS, I can code features very fast. (that's especially important for my side projects).
Besides, Node.js is very lightweight and has a very low footprint at runtime.

We could probably say the same about Python or Go... so it's more a personal preference :)

**Comment by siredwin** [comment]
It seemed like most docker related contributors use Golang mostly.
One more question, will this work with private docker-hub repos? 
I eventually will use ECR and private docker hub.

**Comment by fmartinou** [comment]
Actually, this app is mainly dealing with the official [Docker Registry V2 API](https://docs.docker.com/registry/spec/api/).
So if the underlying system exposes it (docker hub, ECR...), it should work... except for the authentication part which is not very well standardized.
So yes, it will certainly work with private Docker Hub repositories... as soon as I implement the Authentication part ;)

**Comment by fmartinou** [comment]
Starting from version `2.1.0`, WUD can now work with Docker Hub **Private** repositories too ([see the documentation here](https://fmartinou.github.io/whats-up-docker/#/watchers/docker/hub)).

Regards

**Comment by siredwin** [comment]
Thanks.
It seems to be working.
I hope in the future you will add a manual update/restart button for each image on the ui.

**Comment by fmartinou** [comment]
"manual update/restart button for each image on the ui" -> It's planned.

**Comment by fmartinou** [comment]
Hi,

The "manual update/restart button for each image on the ui" is implemented starting from version `2.2.0`
([See UI](https://fmartinou.github.io/whats-up-docker/#/ui/images) & [API](https://fmartinou.github.io/whats-up-docker/#/api/image/watch-image))

**Comment by siredwin** [comment]
Do you recommend using a version number or latest in production?

**Comment by fmartinou** [comment]
My recommendation?
A release version without hesitation.
Immutability is a key feature for production environments.

**Comment by siredwin** [comment]
I like the fact that this service can watch private docker repo, however, I am a little uneasy about the easily visible credentials on the docker-compose file.
```
"WUD_WATCHER_DOCKER_LOCAL_REGISTRIES_HUB_AUTH_LOGIN": "mysuperorganization",
"WUD_WATCHER_DOCKER_LOCAL_REGISTRIES_HUB_AUTH_PASSWORD": "fb4d5db9-e64d-3648-8846-74d0846e55de"
```
What if it was base64 encoded like docker does on config.json i.e `"auth": "base64(USERNAME:PASSWORD)"` which results in `"auth": "VVNFUk5BTUU6UEFTU1dPUkQ="`?
I know it is a small step, but it would look safer especially if I have to share the projects.
Thanks for your project.

**Comment by fmartinou** [comment]
It's very common to inject secrets as env vars in Docker containers.
But I agree; it implies that you can apply some security measures upstream.

_" it would look safer especially if I have to share the projects"_
Be aware that Base64 won't protect you from your password to leak; it would just be a protection against "prying eyes".

Anyway, I get your point and I propose to implement both approaches.

For example, you could either inject (as now):
```
"WUD_WATCHER_DOCKER_LOCAL_REGISTRIES_HUB_AUTH_LOGIN": "mysuperorganization",
"WUD_WATCHER_DOCKER_LOCAL_REGISTRIES_HUB_AUTH_PASSWORD": "fb4d5db9-e64d-3648-8846-74d0846e55de"
```

or inject:
```
"WUD_WATCHER_DOCKER_LOCAL_REGISTRIES_HUB_AUTH_PASSWORD": "bXlzdXBlcm9yZ2FuaXphdGlvbjpmYjRkNWRiOS1lNjRkLTM2NDgtODg0Ni03NGQwODQ2ZTU1ZGU="
```
(if the login is empty, WUD could try to Base64 Decode the password to keep the real login + password)

Would that suit you?

**Comment by siredwin** [comment]
That should work. 
All my servers are going to be shared with third parties. But the repositories need to be private.
Thanks.

**Comment by fmartinou** [comment]
With version `2.3.0`, wud can now support Base64 encoded login:password for Docker Hub:
https://fmartinou.github.io/whats-up-docker/#/watchers/docker/hub

Regards

**Comment by siredwin** [comment]
Perfect, it is working.

**Comment by fmartinou** [comment]
Hi,
I've just released the version `3.0.0` which can now support ECR repositories... which was your initial request ;)

__Please pay attention that some configuration variables have been changed.__
So review your current configuration, compare with the documentation and make the adjustments if needed before upgrading.

Regards.

**Comment by siredwin** [comment]
Thanks, the current implementation works just fine. It might be a while since I get back to monitoring ECR
